File: day-04/python/part2.py
#!/usr/bin/env python3

import sys
import re

required_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]  # "cid" not needed as it is optional
valid_ecl = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
valid_units = ["cm", "in"]

with open(sys.argv[1], 'r') as f:
    passports = f.read().split("\n\n")

valid = 0

for field in passports:
    d = {}
    for b in field.split():
        k, v = b.split(":")
        d[k] = v

    try:
        if not (1920 <= int(d["byr"]) <= 2002):
            continue
        if not (2010 <= int(d["iyr"]) <= 2020):
            continue
        if not (2020 <= int(d["eyr"]) <= 2030):
            continue

        if d["hgt"][-2:] not in valid_units:
            continue

        hgt = int(d["hgt"][:-2])

        if (d["hgt"][-2:] == "cm") and not (150 <= hgt <= 193):
            continue
        if (d["hgt"][-2:] == "in") and not (59 <= hgt <= 76):
            continue

        if not re.match(r"#([0-9a-f]{6})", d["hcl"]):
            continue

        if d["ecl"] not in valid_ecl:
            continue

        if len(d["pid"]) != 9:
            continue
    except:
        continue

    valid += 1

print(valid)

File: day-04/nim/part2.nim
from os import paramStr
import tables
import strutils
import nre except toSeq

let
  passports = paramStr(1).string.readfile.split("\n\n")
  valid_ecl = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
  valid_units = ["cm", "in"]

var valid: int = 0

for field in passports:
  var d = initTable[string, string]()
  for b in field.split():
    var kv = b.split(":")
    d[kv[0]] = kv[1]

  try:
    if parseInt(d["byr"]) notin 1920..2002: continue
    if parseInt(d["iyr"]) notin 2010..2020: continue
    if parseInt(d["eyr"]) notin 2020..2030: continue

    var
      hgt = d["hgt"]
      unit = hgt[^2..^1]
      num = parseInt(hgt[0..^3])

    if unit notin valid_units:
      continue

    if (unit == "cm") and (num notin 150..193):
      continue

    if (unit == "in") and (num notin 59..76):
      continue

    if nre.match(d["hcl"], re"#([0-9a-f]{6})").isNone or len(d["hcl"]) != 7:
      continue

    if d["ecl"] notin valid_ecl:
      continue

    if nre.match(d["pid"], re"[0-9]{9}").isNone or len(d["pid"]) != 9:
      continue
  except:
    continue

  inc(valid)

echo valid
